File: app/src/main/java/com/github/flickr/dependency/downloader/GlideImageDownloader.java
package com.github.flickr.dependency.downloader;


import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;

import javax.inject.Inject;

import rx.Observable;
import rx.subjects.AsyncSubject;

public class GlideImageDownloader implements ImageDownloader {

    @NonNull
    RequestManager glide;

    @Inject
    GlideImageDownloader(@NonNull RequestManager glide) {
        this.glide = glide;
    }

    @Override
    public @NonNull
    Observable<BitmapHolder> downloadBitmap(@NonNull String url) {
        final AsyncSubject<BitmapHolder> notifier = AsyncSubject.create();

        glide.asBitmap().load(url).into(new SimpleTarget<Bitmap>() {
            @Override
            public void onResourceReady(@NonNull Bitmap resource,
                                        @Nullable Transition<? super Bitmap> transition) {
                notifier.onNext(new BitmapHolderImpl(resource));
                notifier.onCompleted();
            }

            @Override
            public void onLoadFailed(@Nullable Drawable errorDrawable) {
                notifier.onError(new Throwable("Can't load image"));
                notifier.onCompleted();
            }
        });

        return notifier;
    }
}

File: app/src/main/java/com/github/flickr/dependency/downloader/GlideImageDownloader.kt
package com.github.flickr.dependency.downloader


import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import com.bumptech.glide.RequestManager
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import rx.Observable
import rx.subjects.AsyncSubject
import javax.inject.Inject

class GlideImageDownloader @Inject constructor(private val glide: RequestManager) : ImageDownloader {

    override fun downloadBitmap(url: String): Observable<BitmapHolder> {
        val notifier = AsyncSubject.create<BitmapHolder>()

        glide.asBitmap().load(url).into(object : SimpleTarget<Bitmap>() {
            override fun onResourceReady(resource: Bitmap,
                                         transition: Transition<in Bitmap>?) {
                notifier.onNext(BitmapHolderImpl(resource))
                notifier.onCompleted()
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                notifier.onError(Throwable("Can't load image"))
                notifier.onCompleted()
            }
        })

        return notifier
    }
}
